ENH - Modify Dataframe.select_dtypes to accept scalar values

Ffisegydd opened the pull request:
This commit relates to GH16855. It allows the Dataframe.select_dtypes
function to accept scalar values as well as list-like values. As such,
it should maintain backwards compatibility.

 - [x ] closes #16855
 - [x ] tests added / passed
 - [x ] passes ``git diff upstream/master --name-only -- '*.py' | flake8 --diff`` (On Windows, ``git diff upstream/master -u -- "*.py" | flake8 --diff`` might work as an alternative.)
 - [x ] whatsnew entry

Commit:
ENH - Modify Dataframe.select_dtypes to accept scalar values

This commit related to GH16855. It allows the Dataframe.select_dtypes
function to accept scalar values as well as list-like values. As such,
it should maintain backwards compatibility.

jreback commented on doc/source/whatsnew/v0.21.0.txt:
```diff
@@ -39,6 +39,7 @@ Other Enhancements
 - :func:`read_feather` has gained the ``nthreads`` parameter for multi-threaded operations (:issue:`16359`)
```
a reason you changed style.ipynb?

Ffisegydd replied:
It got changed automatically by something as part of trying to build docs. I found it actually got deleted at some point and so reverted it from previous commits but the space was different for some reason.

jreback commented on pandas/core/frame.py:
```diff
@@ -2348,10 +2346,15 @@ def select_dtypes(self, include=None, exclude=None):
-            raise TypeError('include and exclude must both be non-string'
-                            ' sequences')
+
+        # GH16855 - If either include or exclude is a non-None scalar then
```
comment not needed

jreback commented on pandas/core/frame.py:
```diff
@@ -2348,10 +2346,15 @@ def select_dtypes(self, include=None, exclude=None):
+        # GH16855 - If either include or exclude is a non-None scalar then
+        # convert to a tuple of length 1 and continue.
+        # This allows, for example, df.select_dtypes(include='object').
+        if not is_list_like(include):
```
use

```
 if include is not None and not is_list_like(include):
    include = [include]
```
and similar

Ffisegydd replied:
Using this logic doesn't cover the case when `include` is `None` as it then gets passed through as `None`, not `[None]`, and causes a `TypeError` later on.

jreback commented on pandas/tests/frame/test_dtypes.py:
```diff
@@ -149,6 +149,10 @@ def test_select_dtypes_include(self):
         ei = df[['k']]
         assert_frame_equal(ri, ei)
 
```
add as a separate test, try both with include and exclude as scalaras (separately)

jreback commented on pandas/tests/frame/test_dtypes.py:
```diff
@@ -162,6 +166,10 @@ def test_select_dtypes_exclude(self):
         ee = df[['a', 'e']]
         assert_frame_equal(re, ee)
 
+        re = df.select_dtypes(exclude=np.number)
```
add this test in there

jreback reviewed (changes requested):
pretty good. some comments.

Commit:
Changes according to comments

Includes:
 - Adding another example to the docstring
 - Separated out the tests into explicit functions to cover the
   different scenario.

 Doesn't implement the change in logic suggested as it doesn't work
 as suggested. Happy to discuss other logic changes as necessary.

Ffisegydd commented:
As discussed in commit message. Implemented all changes apart from the logic change as it misses a necessary edge case. Happy to discuss it though.

jreback commented:
nice PR!

keep em coming!
